packages/ra-core/src/form/FormDataConsumer.tsx

```diff
@@ -64,34 +65,36 @@ export const FormDataConsumerView = <
 >(
     props: Props<TFieldValues>
 ) => {
-    const { children, form, formData, source, index, ...rest } = props;
+    const { children, form, formData, source, ...rest } = props;
     let ret;
 
+    const sourceContext = useSourceContext();
+    // Passes an empty string here as we don't have the children sources and we just want to know if we are in an iterator
+    const matches = ArraySourceRegex.exec(
+        sourceContext?.getSource('') ?? source
+    );
+
     // If we have an index, we are in an iterator like component (such as the SimpleFormIterator)
-    if (typeof index !== 'undefined' && source) {
-        const scopedFormData = get(formData, source);
-        const getSource = (scopedSource: string) => `${source}.${scopedSource}`;
-        ret = children({ formData, scopedFormData, getSource, ...rest });
+    if (matches) {
+        const scopedFormData = get(formData, matches[0]);
+        ret = children({ formData, scopedFormData, ...rest });
```

Comment:
I'm thinking that we could extract the logic for getting scopedFormData in a custom hook depending only on `useSourceContext`. That way, we could encourage the good practice of using this hook instead of expecting the scoped form data to be injected (which is very react-admin v3).

Reply:
a render prop does not inject anything (that's actually one of its strength) and saves you the trouble to declare a component just to call a hook. IMO that's way more pratical. We could do both though to accommodate all use cases.

Reply:
Can I address this in a dedicated PR?